pkg/internal/tests/cross-tests/create.go

```diff
@@ -0,0 +1,119 @@
+)
+
+// Create validates that a Terraform provider witnesses the same input when:
+// - invoked directly with HCL on tfConfig
+// - bridged and invoked via Pulumi YAML on puConfig
+//
+// Create only applies to resources defined with github.com/hashicorp/terraform-plugin-sdk/v2. For cross-tests
+// on Plugin Framework based resources, see
+// github.com/pulumi/pulumi-terraform-bridge/pkg/pf/tests/internal/cross-tests.
+//
+// Create *does not* verify the outputs of the resource, only that the provider witnessed the same inputs.
+func Create(
+	t T, resource map[string]*schema.Schema, tfConfig cty.Value, puConfig resource.PropertyMap,
+	options ...CreateOption,
+) {
+	var opts createOpts
+	for _, f := range options {
+		f(&opts)
+	}
+
+	type result struct {
+		data   *schema.ResourceData
+		meta   any
+		wasSet bool
+	}
+	var tfResult, puResult result
+
+	makeResource := func(writeTo *result) *schema.Resource {
+		return &schema.Resource{
+			Schema: resource,
+			CreateContext: func(_ context.Context, rd *schema.ResourceData, meta any) diag.Diagnostics {
+				*writeTo = result{rd, meta, true}
+				rd.SetId("someid") // CreateContext must pick an ID
+				return nil
+			},
+		}
+	}
+
+	tfwd := t.TempDir()
+	tfd := newTFResDriver(t, tfwd, defProviderShortName, defRtype, makeResource(&tfResult))
+	tfd.writePlanApply(t, resource, defRtype, "example", tfConfig, lifecycleArgs{})
+
+	require.True(t, tfResult.wasSet, "terraform test result was not set")
+
+	resMap := map[string]*schema.Resource{defRtype: makeResource(&puResult)}
+	tfp := &schema.Provider{ResourcesMap: resMap}
+	bridgedProvider := pulcheck.BridgedProvider(
+		t, defProviderShortName, tfp,
+		pulcheck.WithResourceInfo(map[string]*info.Resource{defRtype: opts.resourceInfo}),
+	)
+	pd := &pulumiDriver{
+		name:                defProviderShortName,
+		pulumiResourceToken: defRtoken,
+		tfResourceName:      defRtype,
+	}
+	yamlProgram := pd.generateYAML(t, puConfig)
+
+	pt := pulcheck.PulCheck(t, bridgedProvider, string(yamlProgram))
+
+	pt.Up(t)
+
+	require.True(t, puResult.wasSet, "pulumi test was not set")
+
+	// Compare the result
+
+	assert.Equal(t, tfResult.meta, puResult.meta,
+		"assert that both providers were configured with the same provider metadata")
+
+	// We are unable to assert that both providers were configured with the exact same
+	// data. Type information doesn't line up in the simple case. This just doesn't work:
+	//
+	//	assert.Equal(t, tfResult.data, puResult.data)
+	//
+	// We make due by comparing raw data.
+	assertCtyValEqual(t, "RawConfig", tfResult.data.GetRawConfig(), puResult.data.GetRawConfig())
+	assertCtyValEqual(t, "RawPlan", tfResult.data.GetRawPlan(), puResult.data.GetRawPlan())
+	assertCtyValEqual(t, "RawState", tfResult.data.GetRawState(), puResult.data.GetRawState())
+}
+
+type createOpts struct {
+	resourceInfo *info.Resource
+}
+
+// An option that can be used to customize [Create].
+type CreateOption func(*createOpts)
```

Comment:
For internal code I wonder sometimes if plain struct options are just easier to work with and discover than functional options. That's a stylistic nit though. It's all good.